Feature/indexer tests

pacmessica opened the pull request:
[notion](https://www.notion.so/figmentnetworks/b47c0f8b935741d8ac28717ddeea2038?v=5e4e1d3854094b788ae2b612329f27e8&p=1494c2a0cfbc4089b09ffc1f87da670b)

1. Adds unit tests for tasks in indexer package
2.  Abstracts store to its own interface (`store.Store`) and moves Postgres store implementation to its own package (`store/psql`). The rest of the repo now relies only on the interface rather than a db instance. This will make it much easier in the future if we choose to write integration tests or even switch dbs :)

Note- still missing some tests, created new ticket [here](https://www.notion.so/figmentnetworks/b47c0f8b935741d8ac28717ddeea2038?v=5e4e1d3854094b788ae2b612329f27e8&p=8418199ec0da48599d37fadd73c2bcb0) for remaining tests

pacmessica commented on model/validator_agg.go:
```diff
@@ -27,4 +27,8 @@ func (s *ValidatorAgg) Equal(m ValidatorAgg) bool {
+
+	s.RecentAsValidatorHeight = u.RecentAsValidatorHeight
+	s.AccumulatedUptime = u.AccumulatedUptime
+	s.AccumulatedUptimeCount = u.AccumulatedUptimeCount
```
need these otherwise this doesn't work (from aggregator_tasks.go)

```
			validator := &model.ValidatorAgg{
				Aggregate: &model.Aggregate{
					RecentAtHeight: payload.Syncable.Height,
					RecentAt:       payload.Syncable.Time,
				},

				RecentAsValidatorHeight: payload.Syncable.Height,
			}

			if payload.Syncable.LastInSession {
				if validatorData.Performance.GetOnline() {
					validator.AccumulatedUptime = existing.AccumulatedUptime + 1
				} else {
					validator.AccumulatedUptime = existing.AccumulatedUptime
				}
				validator.AccumulatedUptimeCount = existing.AccumulatedUptimeCount + 1
			}  else {
				validator.AccumulatedUptime = existing.AccumulatedUptime
				validator.AccumulatedUptimeCount = existing.AccumulatedUptimeCount
			}

			existing.Update(validator)
```

silentlight reviewed (commented):
@pacmessica I like your refactoring. It separates interfaces from implementation. Well done!

I am thinking about those `db` attributes. Some of them are store.Store and some of them are specific stores ie. store.ValidatorAgg. I think we should make a decision if we want to always pass store.Store or we pass individual stores and just instead of calling them db we call them ie. validatorAggDb. I think I would prefer option #2 since this way we know what underlying method uses. And if we pass too many different store instances it might be a good indicator that our method is doing too much. Let me know what your thoughts are :)?

pacmessica commented:
> @pacmessica I like your refactoring. It separates interfaces from implementation. Well done!
> 
> I am thinking about those `db` attributes. Some of them are store.Store and some of them are specific stores ie. store.ValidatorAgg. I think we should make a decision if we want to always pass store.Store or we pass individual stores and just instead of calling them db we call them ie. validatorAggDb. I think I would prefer option #2 since this way we know what underlying method uses. And if we pass too many different store instances it might be a good indicator that our method is doing too much. Let me know what your thoughts are :)?

I think being explicit with what you're passing around is good practice, so I also prefer #2. I think the usecase package will look a lot cleaner. Good suggestion!

pacmessica commented on store/psql/store.go:
```diff
@@ -0,0 +1,158 @@
+)
+
+// Store handles all database operations
+type Store struct {
```
psql.Store implements the store.Store interface

silentlight replied:
Ah right!

silentlight commented on indexer/persistor_tasks.go:
```diff
@@ -184,13 +185,13 @@ func (t *validatorAggPersistorTask) Run(ctx context.Context, p pipeline.Payload)
 
 	for _, aggregate := range payload.NewValidatorAggregates {
-		if err := t.db.ValidatorAgg.Create(&aggregate); err != nil {
+		if err := t.db.Create(&aggregate); err != nil {
```
I don't know about you but use of `db` made me confused few times and I went looking for what we actually persisting here since for instance `t.db.Create(&aggregate)` does not really tell me what are we creating here. I know it is for an aggregator but idk which one. So I think we should me it more clear by either:
1. Using something like `validatorAggDb` or `ValidatorAggStore` instead of `db`. This option does not force us to change existing struct attribute when we add new store to it
2. Change the name of `aggregate` to `validatorAgg`

What do you think? I think #1 should be easier to implement

pacmessica replied:
ah yeah! #1 for sure. That's the pattern I'm using elsewhere to distinguish between multiple stores

silentlight reviewed (commented):
@pacmessica looks good to me! Just left one comment about naming ;)

lukanus commented on usecase/worker_handlers.go:
```diff
@@ -1,23 +1,65 @@
+		return fmt.Errorf("WorkerHandlerParams cannot be nil")
+	}
+
+	v := reflect.ValueOf(*b)
```
IMHO Reflection should be really the last resort. Wouldn't it be better just to check for each one like : 
```golang
if b.BlockSummaryDb  == nil { 
    return fmt.Errorf("BlockSummaryDb is required")
} else if !b.BlockSummaryDb.Verify() // opportunity to make some extra validation {
   /// other error 
} else {
   /// profit! :)
}
```

pacmessica replied:
I agree about reflect. Thanks for calling me out 😝 

I removed verification- I think it was necessary if I'm passing the params in a struct because then it could be possible to accidentally forget/delete a param and the compiler wouldn't complain. But I want to switch it back to declaring each param deliberately like this:

```
httpHandlers, err := usecase.NewHttpHandlers(client, db.GetAccounts(), db.GetBlocks(), db.GetEvents(), db.GetSyncables(), db.GetValidators())
```

lukanus commented on usecase/worker_handlers.go:
```diff
@@ -1,23 +1,65 @@
 }
+
+func (b *WorkerHandlerParams) validate() error {
+	if b == nil {
```
If this one is internal method for pointer receiver, you would get nil pointer exception before reaching that point

lukanus commented on usecase/cmd_handlers.go:
```diff
@@ -1,21 +1,49 @@
+		return nil, err
 	}
+
+	return &CmdHandlers{
```
I doubt we benefit here from mapping params to one big structure, just to assign that later. It might be better to just construct the message. Also we are not using this structure anywhere else, just to create CmdHandlers. If there is no additional logic -> I would probably copy paste that to /cli/cmd.go because it is not used in any other context anyway.

pacmessica replied:
I simplified the params and got rid of the `HandlerParams` struct, so I think this isn't an issue anymore. :)

Thanks for all your comments! It was very helpful. Would really appreciate some more feedback on the last couple commits, and then I can clean everything up :)

I don't think it needs to be perfect and super optimized at this time (like how you suggested passing the `FindMostRecenter` interface because the other methods on syncable weren't being used- it's a great suggestion, but I think it would take a while to implement this properly for every method).  As long as the current structure allows for these kinds of improvements easily, then I think it's good enough.

pacmessica replied:
Main goal was to remove the dependence on the psql instance

lukanus reviewed (commented):
Very nice! Nice step toward separation of concerns! 
Left few comments. I think that it may benefit from narrowing interface range or grouping it by the target. The goal would be to reduce number of interfaces as well as shrunk the responsibility of one. 
Eg. io.Read it may be used passing strings.Reader, bytes.Buffer, or json.Encode that have all the other functions. but we don't need to know about all, just this particular one

pacmessica commented on store/validator_era_seq.go:
```diff
@@ -0,0 +1,44 @@
+	Summarize(interval types.SummaryInterval, activityPeriods []ActivityPeriodRow) ([]ValidatorEraSeqSummary, error)
+}
+
+type ValidatorEraSeqSummary struct {
```
good idea! moved it to models package along with some other summaries

pacmessica commented on utils/logger/logger.go:
```diff
@@ -30,49 +31,47 @@ func Init(cfg *config.Config) (*Logger, error) {
-	Log.log.Info(msg, tags...)
-	Log.log.Sync()
+	Log.Info(msg, tags...)
+	Log.Sync()
```
ooo I'll make a separate ticket for that. I think we also need to make that change in oasis

pacmessica replied:
[new ticket](https://www.notion.so/figmentnetworks/b47c0f8b935741d8ac28717ddeea2038?v=5e4e1d3854094b788ae2b612329f27e8&p=927bf9a859e74a319350531db190d67c)

pacmessica commented on store/store.go:
```diff
@@ -1,97 +1,28 @@
+	Getters
+}
+
+type Getters interface {
```
I decided to get rid of the store.Store and store.Getters interface and just use the postgres instance directly in the cli package. 

I liked the idea of grouping though, so I grouped similar stores together. I think it's a lot easier to read and there's fewer params to pass around which is nice. I implemented it for the httphandler [here](https://github.com/figment-networks/polkadothub-indexer/pull/8/commits/5e21fc6879c030aa94ea452a57fc082a37f4b0d6). What do you think?

lukanus replied:
So if you look at it. 
This: 
```httpHandlers, err := usecase.NewHttpHandlers(client, db.GetAccounts(), db.GetBlocks(), db.GetEvents(), db.GetSyncables(), db.GetValidators())```
Is roughly this: 
```httpHandlers, err := usecase.NewHttpHandlers(client, db)```
:) 

Is there any case where accounts would be from different database than events? 
Probably not, so you can as well pass it as one and deal with the separation inside. 
This not is a constructor really, so you don't need to pass this as an interface, and noone would get hurt ;)

pacmessica commented on usecase/account/get_by_height_http_handler.go:
```diff
@@ -1,30 +1,32 @@
 }
 
-func NewGetByHeightHttpHandler(db *store.Store, c *client.Client) *getByHeightHttpHandler {
+func NewGetByHeightHttpHandler(c *client.Client, syncablesDb store.Syncables) *getByHeightHttpHandler {
```
tried that out [here](https://github.com/figment-networks/polkadothub-indexer/pull/8/commits/a4e0e1633937501115989be72722a7f067f008d9), is this what you mean?

lukanus replied:
@pacmessica What you did with this interface `FindMostRecenter` is correct approach. 
This should grow naturally in case it would need more functions underneath. So, if you envision that you may need a wider interface soon other name might be better. But starting from reasonably narrow interface is much better! :)

eesmerdag commented on store/psql/store.go:
```diff
@@ -0,0 +1,191 @@
+	"github.com/figment-networks/polkadothub-indexer/store"
+	"github.com/figment-networks/polkadothub-indexer/types"
+	"github.com/jinzhu/gorm"
+	_ "github.com/jinzhu/gorm/dialects/postgres"
```
why do we need this line here? because of an compile error ?

lukanus replied:
This is how you initialize postgres driver in gorm, because gorm just by itself doesn't have capability of connecting to psql

pacmessica replied:
@lukanus you respond to emir but not me 😢

pacmessica replied:
(I have some questions about some of your comments, could use some quick feedback)

eesmerdag commented on store/psql/store.go:
```diff
@@ -0,0 +1,191 @@
+
+// Close closes the database connection
+func (s *Store) Close() error {
+	return s.db.Close()
```
i think error handling should be implemented or return error should be erased.

pacmessica replied:
what should we do if there's an error?

eesmerdag replied:
i meant when there is an error we don't handle it. but erasing error from method signature ruins interface. i can resolve it for now

lukanus commented on indexer/persistor_tasks_test.go:
```diff
@@ -0,0 +1,517 @@
+	}
+
+	for _, tt := range tests {
+		tt := tt
```
Yes. t.Parallel() are causing this to be necessary.

pacmessica commented on indexer/aggregator_tasks.go:
```diff
@@ -92,6 +93,9 @@ func (t *validatorAggCreatorTask) Run(ctx context.Context, p pipeline.Payload) e
 					validator.AccumulatedUptime = existing.AccumulatedUptime
 				}
 				validator.AccumulatedUptimeCount = existing.AccumulatedUptimeCount + 1
+			} else {
```
which error are you talking about?

eesmerdag replied:
this method is too big i guess we can implement smaller private functions and call them.

`if {
  call method 1
} else {
  call method 2 
  if {  
    call method 2 
  } else { 
    call method 3
 }
}
 `

lukanus replied:
@eesmerdag for me adding separate functions for simple assignation would be huge overkill. 
If we plan to refactor this whole thing @pacmessica I would however do something like this, because there are things we don't really need in this whole expression 

```go

	for stashAccount, validatorData := range parsedValidators {
		existing, err := t.validatorAggDb.FindAggByStashAccount(stashAccount)
		if err != nil {
			if err != store.ErrNotFound {
				return err
			}

			// Create new
			validator := model.ValidatorAgg{
				Aggregate: &model.Aggregate{
					StartedAtHeight: payload.Syncable.Height,
					StartedAt:       payload.Syncable.Time,
					RecentAtHeight:  payload.Syncable.Height,
					RecentAt:        payload.Syncable.Time,
				},

				StashAccount:            stashAccount,
				RecentAsValidatorHeight: payload.Syncable.Height,
			}

			if payload.Syncable.LastInSession {
				if validatorData.Performance.GetOnline() {
					validator.AccumulatedUptime = 1
				}
				validator.AccumulatedUptimeCount = 1
			}

			newValidatorAggs = append(newValidatorAggs, validator)
			continue
		} 


		// Update
		validator := &model.ValidatorAgg{
			Aggregate: &model.Aggregate{
				RecentAtHeight: payload.Syncable.Height,
				RecentAt:       payload.Syncable.Time,
			},

			RecentAsValidatorHeight: payload.Syncable.Height,

			AccumulatedUptime: existing.AccumulatedUptime,
			AccumulatedUptimeCount: existing.AccumulatedUptimeCount,
		}

		if payload.Syncable.LastInSession {
			if validatorData.Performance.GetOnline() {
				validator.AccumulatedUptime++
			}
			validator.AccumulatedUptimeCount++
		} 

		existing.Update(validator)

		updatedValidatorAggs = append(updatedValidatorAggs, *existing)
	
	}
```


```

```

pacmessica replied:
@lukanus this is so much cleaner!! I made the change, thanks 👍 👍

lukanus replied:
Just make sure it is really the same @pacmessica xD , I was doing that in github copy pasting stuff. it should be but check it after me please :)

pacmessica replied:
lol I didn't copy and paste blindly, don't worry. It just happens that it was already perfect so I didn't need to change anything :p

pacmessica commented on indexer/aggregator_tasks_test.go:
```diff
@@ -0,0 +1,347 @@
+	"github.com/pkg/errors"
+)
+
+func TestValidatorAggCreatorTask_Run(t *testing.T) {
```
Not sure what you mean. Can you give an example of an alternative?

eesmerdag replied:
`TestValidatorAggCreatorTask_Run` should be  `RunValidatorAggCreatorTaskTest` ?

lukanus replied:
This has to be TestWhatever because this is how go takes functions in it's test suite https://golang.org/pkg/testing/

pacmessica replied:
@eesmerdag "bottom line" = "underscore" :)

I'm going to leave this as is for now because it's cosmetic, and we've already established this naming convention in all the tests in this package plus oasis so I will side on consistency.

eesmerdag replied:
👍

pacmessica commented on indexer/fetcher_tasks_test.go:
```diff
@@ -0,0 +1,109 @@
+
+		pl := &payload{CurrentHeight: 20}
+
+		mockClient.EXPECT().GetAll(pl.CurrentHeight).Return(&heightpb.GetAllResponse{
```
moving what exactly? lol I think you need to give me more examples, I'm responding to every comment with "example pls" . I just want to be on the same page ;)

eesmerdag replied:
like getting this heightpb.GetAllResponse from a method generateMock.. etc, as i told you it is not big deal

lukanus replied:
@eesmerdag this is just a structure with all data it needs, I don't see a reason why we should wrap it in any external function

> Pluralitas non est ponenda sine necessitate

eesmerdag replied:
@lukanus i thought it would be more creadible.

pacmessica commented on store/errors.go:
```diff
@@ -0,0 +1,7 @@
+import "errors"
+
+var (
+	ErrNotFound = errors.New("record not found")
```
we only have one error, yes. It's imported in lots of places

pacmessica commented on store/psql/event_seq_queries.go:
```diff
@@ -0,0 +1,5 @@
+package psql
+
+const (
```
nothing! good catch

pacmessica commented on store/psql/store.go:
```diff
@@ -0,0 +1,191 @@
+	return s.db.Close()
+}
+
+// GetAccounts gets accounts
```
yeah.. I know what you mean. Golint really wants you to write comments on every exported method/interface. When the method names are already clear it's sort of redundant 🤷. But Golang supports redundancy https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

eesmerdag replied:
🤔 then resolving :)

pacmessica commented on usecase/block/get_by_height.go:
```diff
@@ -7,20 +7,21 @@ import (
 func (uc *getByHeightUseCase) Execute(height *int64) (*DetailsView, error) {
 	// Get last indexed height
-	mostRecentSynced, err := uc.db.Syncables.FindMostRecent()
+	mostRecentSynced, err := uc.syncablesDb.FindMostRecent()
```
created that into a separate ticket [here](https://www.notion.so/figmentnetworks/b47c0f8b935741d8ac28717ddeea2038?v=5e4e1d3854094b788ae2b612329f27e8&p=13d41c665be144149cfbf27819111112)

pacmessica commented:
Thanks for everyone's feedback! I believe I addressed most of the issues you guys brought up, and for others I created new tickets under 'improvement ideas'. Given the size of this PR and the amount of ongoing changes going on in this repo, I think it'd be nice to merge this as it is now (unless there's something that of course must be changed). Would be nice to wrap up the open discussions and get some approvals :)

@silentlight @eesmerdag @lukanus

lukanus commented:
I think this PR waited for too long to be merged. Really Great Job @pacmessica!  
I would vote for merging that regardless of any issues we have with this right now, since this one only brings goodness to our lives. @silentlight , @eesmerdag  what do you think? If there are any further issues we can deal with that in next iterations. 
Also, really sorry it took me so long to reiterate on this pull request @pacmessica :( I promise to improve next time :(

eesmerdag commented on store/blocks.go:
```diff
@@ -0,0 +1,41 @@
+	CreateSeq(*model.BlockSeq) error
+	DeleteSeqOlderThan(purgeThreshold time.Time, activityPeriods []ActivityPeriodRow) (*int64, error)
+	FindSeqByHeight(height int64) (*model.BlockSeq, error)
+	// FindByID(id int64) (*model.BlockSeq, error)
```
i think you forgot to delete this line

eesmerdag reviewed (approved):
just one more comment
